modules/ruxhal/src/platform/aarch64_common/ivc.rs

```diff
@@ -0,0 +1,218 @@
+ 
+         debug!("The memory address of the IVC Info: VA: 0x{:x}, IPA: 0x{:x}", vpa_ivcinfo, pa_ivcinfo);
+ 
+         ivc_hvc_call(HVISOR_HC_IVC_INFO as u32, pa_ivcinfo, HVISOR_HC_IVC_INFO_SIZE);
+         debug!("ivc_hvc_call finished.");
+ 
+         // Safety: At this point we know ivc_info_ptr is valid and allocated
+         let ivc_info: &IvCInfo = unsafe { &*self.ivc_info_ptr };
+ 
+         let control_table_ptr = (ivc_info.ivc_ct_ipas[communication_zone] + __PA as u64) as *mut ControlTable;
+         self.control_table_ptr = control_table_ptr;
+ 
+         info!("IVC Connection established.");
+         Ok(())
+     }
+ 
+     pub fn send_message(&self, message: &str) -> Result<(), &'static str> {
+         if self.ivc_info_ptr.is_null() {
+             return Err("Not connected");
+         }
+     
+         let ivc_info: &IvCInfo = unsafe { &*self.ivc_info_ptr };
+     
+         let control_table: &ControlTable = unsafe { &*self.control_table_ptr };
+     
+         // Suppose we are zone1, we are to send message to zone0.
+         // Therefore use the out_sec_size field of the ControlTable struct (typically 0x1000).
+         let offset = control_table.out_sec_size as u64;
+         let address: u64 = ivc_info.ivc_shmem_ipas[0] + offset + __PA as u64;
+     
+         write_to_address(address, message)?;
+         info!("Message written to shared memory: {}", message);
+     
+         // Modify the control table to inform Zone0 Linux
+         let control_table: &mut ControlTable = unsafe { &mut *self.control_table_ptr };
+         debug!("Ipi_invoke reset to inform Zone0 linux.");
+         control_table.ipi_invoke = 0x0;
+     
+         Ok(())
+     }
+     
+ 
+     pub fn close(&mut self) -> Result<(), &'static str> {
+         if self.ivc_info_ptr.is_null() {
+             return Err("Not connected");
+         }
+         // free the memory
+         let layout = Layout::from_size_align(HVISOR_HC_IVC_INFO_SIZE, HVISOR_HC_IVC_INFO_ALIGN).unwrap();
+         self.allocator.dealloc(unsafe { NonNull::new_unchecked(self.ivc_info_ptr as *mut u8) }, layout);
+         info!("IVC Connection closed.");
+         Ok(())
+     }
+ }
+ 
+ /// ivc "hvc" method call
+ fn ivc_hvc_call(func: u32, arg0: usize, arg1: usize) -> usize {
+     let ret;
+     unsafe {
+         core::arch::asm!(
+             "hvc #4856",
+             inlateout("x0") func as usize => ret,
+             in("x1") arg0,
+             in("x2") arg1,
+             options(nostack)
+         );
+         info!("Ivc call: func: {:x}, arg0: 0x{:x}, arg1: 0x{:x}, ret: 0x{:x}", func, arg0, arg1, ret);
+     }
+     ret
+ }
+ 
+ fn write_to_address(addr: u64, data: &str) -> Result<(), &'static str> {
+     let len = data.len();
+ 
+     if addr == 0 {
+         return Err("Invalid address: 0x0");
+     }
+ 
+     unsafe {
+         let ptr = addr as *mut u8;
+ 
+         for (i, &byte) in data.as_bytes().iter().enumerate() {
```

Comment:
use the `ptr.copy_from`... instead of loop with `for`.

Reply:
fixed.